accord-core/src/main/java/accord/primitives/AbstractKeys.java

```diff
@@ -2,24 +2,22 @@
 
 import java.util.Arrays;
 import java.util.Iterator;
-import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
 import java.util.function.IntFunction;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import accord.api.RoutingKey;
-import accord.utils.IndexedFold;
-import accord.utils.IndexedFoldIntersectToLong;
-import accord.utils.IndexedFoldToLong;
-import accord.utils.IndexedPredicate;
-import accord.utils.IndexedRangeFoldToLong;
-import accord.utils.SortedArrays;
+import accord.utils.*;
 import net.nicoulaj.compilecommand.annotations.Inline;
 
+import static accord.primitives.Sortable.Kind.Key;
+
 @SuppressWarnings("rawtypes")
 // TODO: check that foldl call-sites are inlined and optimised by HotSpot
-public abstract class AbstractKeys<K extends RoutingKey, KS extends AbstractKeys<K, KS>> implements Iterable<K>
+public abstract class AbstractKeys<K extends SortableKey, KS extends Sortables<K, ?>> implements Iterable<K>, Sortables<K, KS>
```

Comment:
```suggestion
// AbstractKeys should be avoided in favor of derived classes because it doesn't restrict to a single kind of key (Routable, Seekable)
public abstract class AbstractKeys<K extends SortableKey, KS extends Sortables<K, ?>> implements Iterable<K>, Sortables<K, KS>
```

Similar comment is also applicable to `Sortable` and `Sortables`.

Reply:
As we discussed offline, sometimes you are willing to accept either